### PR TITLE
[WIP][ApiBundle] Sparse Fieldsets

### PR DESCRIPTION
**Pull request by umpirsky**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | #3912 |
| License | MIT |

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/spec/Serializer/Exclusion/SparseFieldsetsExclusionStrategySpec.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    function it_implements_exclusion_strategy_interface()
+    {
+        $this->shouldImplement(ExclusionStrategyInterface::class);
+    }
+
+    function it_should_never_skip_class(PropertyMetadata $property, Context $navigatorContext)
+    {
+        $this->shouldSkipProperty($property, $navigatorContext)->shouldReturn(false);
```
Comment (umpirsky) [inline src/Sylius/Bundle/ApiBundle/spec/Serializer/Exclusion/SparseFieldsetsExclusionStrategySpec.php]: This is silly:

```
./bin/phpspec run src/Sylius/Bundle/ApiBundle/spec/Serializer
Sylius/Bundle/ApiBundle/Serializer/Exclusion/SparseFieldsetsExclusionStrategy     
  30  - it should never skip class
      no skipProperty([array:2]) matcher found for
      [obj:Sylius\Bundle\ApiBundle\Serializer\Exclusion\SparseFieldsetsExclusionStrategy].

                         66%                                     33%             3
1 specs
3 examples (2 passed, 1 broken)
15ms
```

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 2]
```diff
@@ -107,6 +107,28 @@
     </parameters>
 
     <services>
+        <!-- todo -->
+        <service id="sylius.view_handler.default" class="Sylius\Bundle\CoreBundle\View\ViewHandler" decorates="fos_rest.view_handler.default">
```
Comment (umpirsky) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: @pjedrzejewski What is the best way to override it?

File: src/Sylius/Bundle/ApiBundle/Serializer/Exclusion/SparseFieldsetsExclusionStrategy.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        if (null === $fields) {
+            return false;
+        }
+
+        return !in_array($name, $fields);
+    }
+
+    private function getFields(PropertyMetadata $property, Context $navigatorContext)
+    {
+        $key = $navigatorContext->getMetadataFactory()->getMetadataForClass($property->class)->xmlRootName;
```
Comment (umpirsky) [inline src/Sylius/Bundle/ApiBundle/Serializer/Exclusion/SparseFieldsetsExclusionStrategy.php]: We include by class metadata here. My idea is to do it based on property metadata, but I couldn't find a way to get it.

**Comment by pjedrzejewski** [comment]
I hope you like ApiTestCase. :D We should definitely give more thought about security here because I can imagine this would allow to serialize password field from users. ;) Would be cool to specify which fields are only permitted.

Also, the view handler may not be necessary after #3762 is merged today as it has custom view handler already implemented so we could handle it there. WDYT?

**Comment by umpirsky** [comment]
@pjedrzejewski Yes, it's cool in deed. :)

Good point about security. But specifying which fields are permitted may be lots of work?
Maybe blacklist instead...but still can't think of something elegant.

Also https://github.com/Sylius/Sylius/pull/3938#discussion_r50320665 is something I would like to fix.

Good work on #3762, will wait for this one to get merged first. :+1:

**Comment by nakashu** [comment]
From experience it is worth to specify explicitly which fields are allowed. 
You have all fields exlcuded by default, and only exclude the default set, which the user can make even sorter by defining which field he wants.

Why is it better? Because with the combination of excluding specific fields and serialization groups you will eliminate infinite nested serialization on associated resources.
For example User, Groups, Permission triangle
User belongs to groups and have permission.
Groups have permissions and users.
Permissions have users and groups.

I can provide an code example how it is solved by, exclusion and serialization groups, if needed.

**Comment by umpirsky** [comment]
@nakashu Thanks for your feedback, any example can help so if you can share please do. :+1:

**Comment by nakashu** [comment]
Quick Serialization groups explanation:

 * <resourceName>Detail serializes that property only for single resource retrieval. / getByID
 * <resourceName>List serializes that property only for listed/paginated resource retrieval.

```php
/**
* @ExclusionPolicy("all")
*/
class User implements UserInterface 
{
    /**
     * @Expose
     * @var integer
     */
    private $id;

    // note that password is NOT exposed, ever.
    /**
     * @var string
     */
    private $password;

    /**
     * @Expose
     * @Groups({"userDetail"})
     * @var Group[]|ArrayCollection
     */
    private $groups;

    /**
     * @Expose
     * @Groups({"userDetail", "groupDetail"})
     * @var Privilege[]|ArrayCollection
     */
    private $privileges;
    // ... other properties and methods
}
/**
* @ExclusionPolicy("all")
*/
class Group implements GroupInterface
{
    // ... other properties
    /**
     * @Expose
     * @Groups({"groupDetail", "userDetail"})
     *
     * @var Privilege[]|ArrayCollection
     */
    private $privileges;

    /**
     * @Expose
     * @Groups({"groupDetail"})
     *
     * @var User[]|ArrayCollection
     */
    private $users;
    // ... other properties and methods
}

 * @ExclusionPolicy("all")
 */
class Privilege implements PrivilegeInterface
{
    /**
     * @var User[]|ArrayCollection
     */
    private $users;

    /**
     * @var Group[]|ArrayCollection
     */
    private $groups;
}
```

In the controller it is important to serialize also with the Default groups.
Here it is with annotation, but serialization context can be used easily as well to build the groups dynamicly.
```php
    /**
     * @Annotations\View(serializerGroups={"Default","groupList"})
     */
    public function getGroupIndexAction()
    /**
     * @Annotations\View(serializerGroups={"Default","groupDetail"})
     */
    public function getGroupDetailAction($groupId)
    /**
     * @Annotations\View(serializerGroups={"Default","userDetail"})
     */
    public function getUserDetailAction($groupId)
```

With this setup, you'll eliminated the recursion, and have exact controll what will return where.
I'll describe it in word and json data are bellow

For Group Detail you'll get list of its privileges and users with its own special privileges if assigned.
- on the nested users you WONT get Groups again, as it would create recursion
- on the nested privileges you WONT get Groups and Users again as it would create recursion

For User Detail you'll get list of its privileges and groups with its own privileges.
- on the nested privileges you WONT get Groups and Users again as it would create recursion
- on the nested groups you WONT get Users again, as it would create recursion
- on the nested groups.privileges you WONT get Groups and Users again as it would create recursion

Groups Listing:
```json
{
  "limit": "20",
  "offset": 0,
  "total": 5,
  "pages": 1,
  "page": 1,
  "_links": {
    "self": {
      "href": "/api/v1/groups?limit=20"
    },
    "first": {
      "href": "/api/v1/groups?limit=20"
    },
    "last": {
      "href": "/api/v1/groups?limit=20"
    }
  },
  "_embedded": {
    "items": [
      {
        "id": 1,
        "title": "Admins - Read",
        "handle": "ADMIN_READ",
        "status": 1
      }      
    ]
  }
}
```

Group Detail:
```json
{
  "id": 2,
  "title": "Admins - Write",
  "handle": "ADMIN_WRITE",
  "privileges": {
    "1": {
      "id": 1,
      "title": "admin access",
      "handle": "ROLE_ADMIN_ACCESS"
    },
    "2": {
      "id": 2,
      "title": "users own read",
      "handle": "USERS_OWN_READ"
    },
  },
  "users": [
   {
      "id": 1,
      "email": "email@example.com",
      "status": 1,
      "privileges": {
         "99": {
           "id": 99,
           "title": "special for only this one user privilege",
           "handle": "SPECIAL_PRIVILEGE"
         }
      }
   }
   ]
}
```
User Resource

```json
{
  "id": 1,
  "email": "email@example.com",
  "privileges": {
    "99": { "user special privildge data" } 
  },
  "groups": {
    "2": {
      "id": 2,
      "title": "Admins - Write",
      "handle": "ADMIN_WRITE",
      "privileges": {
        "1": { "permission data" },
        "2": { "permission data" } 
      }
    }
  },
}

**Comment by umpirsky** [comment]
Thanks @nakashu!

But I am just thinking...we never expose passwords or any sensitive data via API, so they will never be included. Right @pjedrzejewski?

**Comment by nakashu** [comment]
@umpirsky 
Only fields that got @Expose() annotation will be serialized, as the class got @ExclusionPolicy("all"), meaning everythings is exluded by default.

Putting @Expose on a property, means you add it to the "Default" serialization group. That's why the group is there in the controller, so properties that dont have other groups (like userDetail), will be included in the output.

btw.: fixed first 3 lines of my example as the part was stripped as I used html tags there.

**Comment by umpirsky** [comment]
@nakashu Sure thing, I just wanted to check with @pjedrzejewski that we never expose sensitive user data to serializer.

**Comment by bendavies** [comment]
@umpirsky are you using this in your project? I might be interested in using it i mine.

**Comment by umpirsky** [comment]
@bendavies Yes, I used to use it on one.

**Comment by bendavies** [comment]
@umpirsky hey, thanks. is it pretty much this exact code or have you improved it?

**Comment by umpirsky** [comment]
@bendavies No, that is pretty much it.

**Comment by pjedrzejewski** [comment]
Closing this because it won't be in the core for now. We will have a look at this for SyliusShopApiPlugin.

**Pull request closed without merging**